std/experimental/perpetual.d

```diff
@@ -0,0 +1,653 @@
+private import std.string;
+private import core.stdc.stdio;
+private import core.stdc.stdlib;
+private import std.stdio;
+private import std.conv;
+
+
+
+version (Windows)
+{
+private import core.sys.windows.windows;
+private import std.utf;
+private import std.windows.syserror;
+}
+else version (Posix)
+{
+private import core.sys.posix.fcntl;
+private import core.sys.posix.unistd;
+private import core.sys.posix.sys.mman;
+private import core.sys.posix.sys.stat;
+}
+else
+{
+    static assert(0);
+}
+
+
+
+package class ShMem
+{
+    enum Mode
+    {
+          readOnly		/// Read existing file
+        , readWrite		/// Read/Write, create if not existing
+        , noExpand		/// Read/Write existing file, do not increaze file size
+    }
+
+
+
+
+
+	this(string path, size_t len, Mode mode =Mode.readWrite)
+	{
+		fileOpen(path, mode);
+		scope(failure) { close(_fd); _fd=-1; }
+
+		auto size=fileSize(path);
+		if(size < len)
+		{
+			enforce(writeable && mode != Mode.noExpand, path~": file is too short");
+			_index=size;
+			_master=true;
+			size=fileExpand(path, len);
+		}
+		else
+		{
+			_index=len;
+		}
+
+		map(path, len);
+	}
+
+
+	this(string path, Mode mode =Mode.readWrite)
+	{
+		fileOpen(path, mode);
+		scope(failure) { close(_fd); _fd=-1; }
+
+		auto len=fileSize(path);
+		enforce(len > 0, path~": zero size file");
+		_index=len;
+		_master=false;
+
+		map(path, len);
+	}
+
+
+	~this()
+	{
+		if(_fd >= 0)
+		{
+			unmap();
+			close(_fd);
+			_fd=-1;
+		}
+	}
+
+
+    void[] opSlice()
+    {
+        return _data[0..$];
+    }
+
+    void[] opSlice(size_t i, size_t k)
+    {
+        return _data[i..k];
+    }
+
+    void opIndex(size_t i)
+    {
+        return _data[i];
+    }
+
+    @property auto length() const { return _data.length; }
+
+    @property bool writeable() const { return _writeable; }
+
+    @property bool master() const { return _master; }
+
+    @property auto unplowed() const { return _index; }
+
+	
+
+
+
+
+
+
+	version(Posix)
+	{
+		private int _fd;
+		private void[] _data;
+		private bool _writeable;
+		private bool _master;
+		private size_t _index;
+	}
+	version(Windows)
+	{
+
+	}
+	
+
+	private int fileOpen(string path, Mode mode)
+	{
+		version(Posix)
+		{
+			/// in read-only mode, the file must exist
+			if(mode == Mode.readOnly)
+			{
+				_fd=core.sys.posix.fcntl.open(toStringz(path), O_RDONLY);
+				errnoEnforce(_fd >= 0, path);
+				_master=false;
+				_writeable=false;
+				return _fd;
+			}
+			else
+			{
+				/// try to create the file
+				_fd=.open(toStringz(path), O_RDWR|O_CREAT|O_EXCL, S_IRUSR|S_IWUSR);
+				if(_fd >= 0)
+				{
+					_master=true;
+					_writeable=true;
+					return _fd;
+				}
+
+				/// otherwise, try to open read/write existing file
+				_fd=.open(toStringz(path), O_RDWR, S_IRUSR|S_IWUSR);
+				if(_fd >= 0)
+				{
+					_master=false;
+					_writeable=true;
+					return _fd;
+				}
+
+				errnoEnforce(false, path);
+			}
+		}
+		else version(Windows)
```

Comment:
Still need to filled. I'm not very good with Windows system calls, neither have a machine to compile and test, so I left it to the last minute and forgot.
--------------------------------------------
On Sat, 9/19/15, Jack Stouffer <notifications@github.com> wrote:

 Subject: Re: [phobos] perpetual: new module for mapping value types into memory (#3625)
 To: "D-Programming-Language/phobos" <phobos@noreply.github.com>
 Cc: "Sergei Degtiarev" <sdegtiarev@yahoo.com>
 Date: Saturday, September 19, 2015, 3:13 PM
 
 In std/experimental/perpetual.d:
 > +					return _fd;
 > +				}
 > +
 > +				/// otherwise, try to open read/write existing
 file
 > +				_fd=.open(toStringz(path), O_RDWR,
 S_IRUSR|S_IWUSR);
 > +				if(_fd >= 0)
 > +				{
 > +					_master=false;
 > +					_writeable=true;
 > +					return _fd;
 > +				}
 > +
 > +				errnoEnforce(false, path);
 > +			}
 > +		}
 > +		else version(Windows)
 
 What's with all the empty
 version(Windows) blocks?
 
 —
 Reply to this email directly or view
 it on GitHub.